frontend/ui/widget/textviewer.lua

```diff
@@ -103,38 +100,13 @@ function TextViewer:init()
         }
     end
 
-    local closeb = CloseButton:new{ window = self, padding_top = Size.padding.tiny, }
-    local title_text = TextBoxWidget:new{
```

Comment:
Note that we were using TextBoxWidget, which allowed multilines. TitleBar does not yet support that.
I think I had seen some multiline titled TextViewer quite often, but can't remember which it was.
Looking at the code, I see only viewing CSS files from View HTML can have a dynamic title, but it feels it was something else.
Anyway, we can probably go without a TextBoxWidget.
If needed, we can either:
- add support for TextBoxWidget (either via an option `multiline=true` or by auto-checking if it is needed)
- checking that if truncated, reduce the title font size so all fits

Reply:
Got it
https://github.com/koreader/koreader/blob/00b08d7b54bc6d0ab563caece98dcce41a128897/frontend/apps/reader/modules/readertypography.lua#L200

![4](https://user-images.githubusercontent.com/62179190/148516481-f4d770fa-6878-4726-9cb6-c8694a579536.png)

Reply:
> Note that we were using TextBoxWidget, which allowed multilines. TitleBar does not yet support that.
I think I had seen some multiline titled TextViewer quite often, but can't remember which it was.

I didn't realize that, but note that this might also be an issue in languages other than English then.

Reply:
> * by auto-checking if it is needed

Think so, as
-smaller font in the title seems not good
-translations might be longer

What are the reasons for not using the TextBoxWidget always?

Reply:
> What are the reasons for not using the TextBoxWidget always?

Because TextBoxWidget is less easy with font metrics, for baseline and vertical centering and alignment with the cross and other icons.

> -smaller font in the title seems not good
-translations might be longer

Was going to suggest it's better to go with a TextWidget with a auto smaller font size :)
I think translations might only be a tad larger, and a smaller font size would still look better: it would be centered and aligned with the cross widget. A multiline TextBoxWidget would make that uglier (align the first line ? The middle of the 2 lines ? A 2nd line with a single word would kill the balance).

Reply:
Here's the longest title I could find at a glance.
![Screenshot_2022-01-07-10-16-25-18_1aff61e6bd981aaa4ba9d0245eb6c04f](https://user-images.githubusercontent.com/202757/148521683-2737a371-172f-4126-8829-8d263396900e.jpg)

Reply:
> a TextWidget with a auto smaller font size :)

Well, maybe, we have it in buttons, so let's have it everywhere.

Reply:
Oh, yes, we have all these titles without a CloseButton too to tweak :/

Current using TextBoxWidget:
![image](https://user-images.githubusercontent.com/24273478/148521938-49d9c70f-b187-4acc-b71e-b4a9d9ba5970.png)
This PR, using font size reduction until it fits:
![image](https://user-images.githubusercontent.com/24273478/148522797-6ac3ff87-5d81-4a40-b115-d110302e959e.png)

Reply:
The first looks better as to me.

Reply:
I prefer the 2nd.
Also, if we go with TitleBar for all widgets, and if using multilines, people using the German UI with its big wording will mostly get 2 lines titlebar everywhere :) having bigger titles with fixed strings, and less height for the interesting content.
May be I made TextViewer use a TextBoxWidget because the dynamic title could have some interesting info, ie the CSS file name in View HTML, but for most other titles, we were just fine with a truncation as it's a fixed string, so people can infer what's truncated.

If we don't want to go with auto-multiline (in these other static contexts), should we:
- have TitleBar truncate per default, and have an option to have it reduce font size for fit
- have TitleBar always reduce font size to fit
- same as above with TextBoxWidget (but I'll hate having to care about the 2 possible different metrics)
- all of the above (with 2 options) :/

Reply:
> * having to care about the 2 possible different metrics

Ah, I remember that vertical_shift in checkbuttons for InputDalog.
Is it possible to fix the baseline forever directly in TBW?

Reply:
There is no baseline to fix in TBW:
- in TBW, we have multiple lines: the padding at top or bottom is the same on all lines and is what makes the interline space (or, the top and bottom padding are made from `line_height = 0.3, -- in em`). I don't want to have to do something different for the 1st and the last line (and doing that could make the text oddly positionned in its container).
- in TW, having a single line, we can have any padding above/below without any side effects, it helps centering the text in its container on influence the container height, in general (in TitleBar, I use padding=0 so I can control the centering myself outside).

The only helper could be adding a `TextBoxWidget:getBaselineShiftFromTextWidget(textwidget_nondefault_padding)` that would instantiate a TextWidget with the same parameters and return the difference, that we could add/reduce in some code that could do as if it only deals with TextWidget.

So, you're inviting me to go with `all of the above (with 2 options) :/` ? :)
(Because in TextViewer for CSS in View HTML, I surely don't want to waste some height with the title, and get as much height for the CSS text below :)
Do you get truncated titles in regular widgets with title, in Russian ?
@zwim @Frenzie : how would you prefer your long DE or NL titles ? Truncated ? Not truncated but as big and multiline ? Not truncated but smaller and single line ?

Reply:
An idea: if the title is truncated, tap on it shows infomessage with full title (as long-press in menu).

Reply:
@poire-z While Dutch, French and German are often a touch longer than English, they're in the same ballpark most of the time. It's really more Russian or something which is full of amazingly long strings.

Reply:
> An idea: if the title is truncated, tap on it shows infomessage with full title (as long-press in menu).

Big meh, unless it's getting excessively long only because of a filename or something. :-)

Reply:
> if the title is truncated, tap on it shows infomessage with full title (as long-press in menu).

One place where we may get long truncated titles is DictQuickLookup with the dict name (ie My Académie Française one in the recent screenshots).
We can't hijack tap, as it may be used (in DictQuickLookup, to set this dict as a preferred dict).
We can't hijack hold, as it can be used if MovableContainer to move the window around (and I use that after View HTML, to move the CSS window outside screen to look again at the HTML, and I usually do that by holding on the title).

Reply:
Well, the tap here could refer to a tap on the `…` area rather than the entire titlebar. (But as I said, meh from me.)

Reply:
Btw, any of these 2 solutions (multiline or smaller size) makes everything complicated if we were to support them when we may be using :setTitle() , like when updating the dict title:
- multiline would need to reshape the whole DictQuickLookup when a title overflows or not, as the title may switch from one line to 2 lines
- space reduction would need to reshape the whole TitleBar, because we can start with a title that does not overflow, and if the updated title does overflow and we reduce its font size, all the paddings have to change...